File: 백준/알고리즘/BFS/1260.py
import sys
from collections import deque
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

n,m,v = map(int, input().split())
graph = [[] for _ in range(n+1)]
visited = [0] * (n+1)

for i in range(m):
  a, b = map(int, input().split())
  graph[a].append(b)
  graph[b].append(a)

for i in range(n+1): graph[i].sort()

def bfs(v):
  q = deque([v])
  visited[v] = 1
  cnt = 2
  while q:
    v = q.popleft()
    print(v, end=' ')
    for k in graph[v]:
      if visited[k] == 0:
        visited[k] = cnt
        cnt += 1
        q.append(k)

cnt = 1
def dfs(v):
  global cnt
  visited[v] = cnt
  print(v, end=' ')
  for k in graph[v]:
    if visited[k] == 0:
      visited[k] = cnt
      cnt += 1
      dfs(k)

dfs(v)
visited = [0] * (n+1)
print()
bfs(v)

"""
import sys
from collections import deque
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

n, m, v = map(int, input().split())
graph = [[0] * (n+1) for _ in range(n+1)]

bfsVisited = [0] * (n+1)
dfsVisited = [0] * (n+1)

dfsCount = 1

for _ in range(m):
  a, b = map(int, input().split())
  graph[a][b] = graph[b][a] = 1

print(graph)

def bfs(v):
  q = deque([v])
  bfsVisited[v] = 1
  while q:
    v = q.popleft()
    print(v, end=' ')
    for i in range(1, n+1):
      if bfsVisited[i] == 0 and graph[v][i] == 1:
        q.append(i)
        bfsVisited[i] = 1

def dfs(v):
  dfsVisited[v] = 1
  print(v, end=' ')
  for i in range(1, n+1):
    if dfsVisited[i] == 0 and graph[v][i] == 1: dfs(i)

dfs(v)
print()
bfs(v)
"""

File: 백준/자바스크립트/그래프/1260.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.front = null;
    this.rear = null;
    this.size = 0;
  }

  isEmpty() {
    return this.size === 0;
  }

  push(data) {
    const newNode = new Node(data);
    if (this.isEmpty()) this.front = newNode;
    else this.rear.next = newNode;
    this.rear = newNode;
    this.size++;
  }

  shift() {
    if (this.isEmpty()) return;
    const data = this.front.data;
    this.front = this.front.next;
    this.size--;
    return data;
  }
}

const fs = require("fs");
const path = process.platform === "linux" ? "/dev/stdin" : "../example.txt";
const input = fs
  .readFileSync(path)
  .toString()
  .trim()
  .split("\n")
  .map((x) => x.split(" ").map(Number));

const [N, M, V] = input[0];
// let graph = new Array(N + 1);
// for (let i = 0; i < graph.length; i++) {
//   graph[i] = [];
// }
let graph = Array.from(Array(N + 1), () => []);

for (let i = 0; i < M; i++) {
  const [a, b] = input[i + 1];
  graph[a].push(b);
  graph[b].push(a);
}

console.log(graph);

graph.forEach((element) => {
  element.sort((a, b) => a - b);
});

let visited = new Array(N + 1).fill(0);
let dfs_result = [];

function DFS(v) {
  if (visited[v]) return;
  visited[v] = 1;
  dfs_result.push(v);
  for (let i = 0; i < graph[v].length; i++) {
    let next = graph[v][i];
    if (visited[next] === 0) {
      DFS(next);
    }
  }
}
DFS(V);
console.log(dfs_result.join(" "));

let bfs_result = [];
visited.fill(0);

function BFS(v) {
  let queue = new Queue();
  queue.push(v);

  while (!queue.isEmpty()) {
    let x = queue.shift();
    if (visited[x] === 1) {
      continue;
    }
    visited[x] = 1;
    bfs_result.push(x);
    for (let i = 0; i < graph[x].length; i++) {
      let next = graph[x][i];
      if (visited[next] === 0) {
        queue.push(next);
      }
    }
  }
}
BFS(V);
console.log(bfs_result.join(" "));
